README
Parchment-proxy
===============

Parchment-proxy is a simple Google App Engine server for proxying Interactive Fiction files for web interpreters like Parchment.

You may install your own, or use ours at http://zcode.appspot.com/proxy/

How to use
==========

Examples:
 - http://zcode.appspot.com/proxy/?url=http://mirror.ifarchive.org/if-archive/games/springthing/2007/Fate.z8
 - http://zcode.appspot.com/proxy/?callback=jsonp12754&encode=base64&url=http://mirror.ifarchive.org/if-archive/games/springthing/2007/Fate.z8

Access the proxy by the /proxy/ URL. The / URL will still work, but has a depreciated API.

Parameters:
 - url
   required, the URL of the story to access
   
 - encode
   set encode=base64 to base64 encode the story file
   
 - callback
   a callback function for JSONP
   If you're using jQuery, set the dataType to 'jsonp' and it will automatically create the callback function and add this parameter for you. Other libraries may do the same.
   If you use a callback, also set encode=base64.

Parchment-proxy will send the data with a Content-Type header of 'text/plain; charset=ISO-8859-1' and an Access-Control-Allow-Origin header of '*' for cross-site AJAX requests. Similarly, it will handle an OPTIONS request if you need to preflight your cross-site requests.

There is currently a limit no limit for requested files, but as big files aren't cached, please be gentle!


CHANGES
Change log started October 2009.

February 2010
================================================================================
Added support for using ETags.

June 2010
================================================================================
Add support for Glulx.
Handle errors properly.
Handle OPTIONS and send Access-Control headers for preflighted requests.
Add support for big file requests, by making urlfetch requests with a Range header.

February 2010
================================================================================
 + A few changes were needed for it to actually work.
 + Add support for a callback function.

January 2010
================================================================================
 + Began working on version 2
 + Split out get_file() to mirror.get().
 + Renamed MainHandler to LegacyHandler and added ProxyHandler.

October 2009
================================================================================
 - Use simplejson so json.py can be removed.
 + Use memcache to cache file downloads.


mirror.py
# Parchment-proxy - A proxy for fetching web data for Parchment
# Copyright 2008-2010 The Parchment-proxy contributors (see CONTRIBUTORS)
# Released under a BSD-like licence, see LICENCE

# mirror.py: File download/retrieval functions
# Will eventually form a mirror of the IF Archive

import logging

from google.appengine.api import memcache
from google.appengine.api import urlfetch

from errors import *

# Maximum size of a story file we'll download, in bytes.
MAX_FILE_SIZE = 1000000

def get(url):
	'''Get a file, using memcache if possible'''
	
	# Check memcache for this url
	data = memcache.get(url)
	if data is not None:
		return data
	
	# Missed! Attempt to download it now
	else:
		result = fetch_single(url)
		data = result.content
		
		# HTTP Status code
		if result.status_code != 200:
			raise ProxyError('got status code %d for url %s' % (result.status_code, url))

		# Check it's a valid glulx/zcode story file, before downloading the rest
		if data.startswith('FORM') or data.startswith('Glul') or ord(data[0]) < 9:
			pass
		else:
			raise ProxyError('url does not contain a story file')
		
		# Check for oversized requests
		if result.content_was_truncated:
			data += fetch_big(url, int(result.headers['Content-Length']), len(data) - 1)
		
		# All good... cache it and return (don't cache big files for now)
		if len(data) <= MAX_FILE_SIZE:
			if not memcache.add(url, data, 86400):
				logging.error('Memcache set failed for url ' + url)
		
		return data

def fetch_single(url, headers={}):
	'''Get one part of a file'''
	try:
		request = urlfetch.fetch(url, headers=headers, allow_truncated=True, deadline=10)
	except urlfetch.DownloadError, e:
		raise ProxyError('could not contact server, or server too slow', e)
		
	return request

def fetch_big(url, size, last):
	'''Get the rest of a big file'''
	# Consider changing later to request the parts async
		
	# Range for the next request
	minrange = last + 1
	maxrange = (last + MAX_FILE_SIZE)
	# Can we trust the Content-Length header? What if it's gzipped?
	if maxrange >= size:
		maxrange = size - 1
	headers = {'Range': 'bytes=%d-%d' % (minrange, maxrange)}
	
	# Make each URL unique, otherwise the cache will only return the first part (What about the headers?!?)
	tempurl = url + '#' + str(minrange)
	
	# Request the next part of the file
	request = fetch_single(tempurl, headers)
	data = request.content
	
	# If that's not all, recursively fetch the next part
	if maxrange < size - 1:
		data += fetch_big(url, size, maxrange)
	
	return data


main.py
#!/usr/bin/env python
# Parchment-proxy - A proxy for fetching web data for Parchment
# Copyright 2008-2011 The Parchment-proxy contributors (see CONTRIBUTORS)
# Released under a BSD-like licence, see LICENCE

# main.py: Main server and response handler

import base64
import hashlib
import logging
import sys
import traceback

from django.utils import simplejson
from google.appengine.api import memcache
from google.appengine.ext import webapp
from google.appengine.ext.webapp.util import run_wsgi_app

from errors import *
import mirror

class ProxyHandler(webapp.RequestHandler):
	'''The parchment-proxy server itself'''
	
	def initialize(self, request, response):
		webapp.RequestHandler.initialize(self, request, response)
		
		# Add our code var
		self.code = 200
	
	def get(self):
		# Parameters
		url = self.request.get('url').replace(' ', '%20')
		callback = self.request.get('callback')
		encode = self.request.get('encode')
		ifnonematch = self.request.headers.get('If-None-Match') or ''
		
		if not url:
			# Show the home page
			self.redirect('/')
			return
		
		# Set headers for allowing cross domain XHR and content type
		self.response.headers['Access-Control-Allow-Origin'] = '*'
		self.response.headers['Content-Type'] = 'text/plain; charset=ISO-8859-1'
		
		# Check for a cached hash
		hash = memcache.get(url + '_hash')
		if hash == ifnonematch:
			self.response.set_status(304)
			self.response.headers['ETag'] = hash
			return
		
		# Get this URL
		data = mirror.get(url)
		
		# Generate the hash if needed
		if not hash:
			hash = '"' + hashlib.md5(data).hexdigest() + '"'
			if not memcache.add(url + '_hash', hash, 86400):
				logging.error('Memcache set failed for hash of url ' + url)
		
		# Base64 encode the data if required
		if encode == 'base64':
			data = base64.b64encode(data)
		
		# Handle a callback function
		if callback:
			# Warning, data must be escaped too
			data = callback + '("' + data + '")'
		
		# Sent the data
		self.response.headers['ETag'] = hash
		self.response.out.write(data)
		
	def options(self):
		# Send Access-Control headers for preflighted requests
		
		self.response.headers['Access-Control-Allow-Origin'] = '*'
		self.response.headers['Access-Control-Allow-Methods'] = 'GET, OPTIONS'
		self.response.headers['Access-Control-Allow-Headers'] = self.request.headers['Access-Control-Request-Headers']
	
	def handle_exception(self, e, debug):
		# Handle exceptions
		self.response.clear()
		
		# Set the HTTP status code
		if self.code == 200:
			self.code = 500
		self.response.set_status(self.code)
		
		# Write out the error message
		self.response.out.write(e)
		
		# Log this error too
		logging.error(repr(e))
		logging.error(''.join(traceback.format_tb(sys.exc_info()[2])))
		

class LegacyHandler(webapp.RequestHandler):
	'''The original jsonp proxy server'''
	def get(self):
		url = self.request.get("url")
		jsonp = self.request.get("jsonp")

		try:
			if not url:
				# Show the home page
				self.print_home()
				return

			url = url.replace(" ", "%20")
			data = mirror.get(url)
			response = {'data': base64.b64encode(data)}
		  
		except Exception, e:
			response = {"error" : repr(e)}

		self.response.headers["Content-Type"] = "text/javascript"
		self.response.out.write("%s(%s);" % (jsonp, simplejson.dumps(response)))
	
	def print_home(self):
		# Print a home page
		self.response.headers["Content-Type"] = "text/html"
		self.response.out.write('''
<!doctype html>
<title>Parchment-proxy</title>
<h1>Parchment-proxy</h1>
<p>This is the proxy for Parchment the web IF interpreter.
<p>If you want to read a story with Parchment go to <a href="http://parchment.toolness.com/">http://parchment.toolness.com/</a>
<p>If you want to know more about Parchment-proxy go to <a href="http://github.com/curiousdannii/parchment-proxy">http://github.com/curiousdannii/parchment-proxy</a>
		''')

def main():
	application = webapp.WSGIApplication([
		('/proxy/?', ProxyHandler),
		('/', LegacyHandler),
	], debug=True)
	run_wsgi_app(application)

if __name__ == '__main__':
	main()
